File: July 2022/July_14.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    map<int, int> mp;
    int n, id;
    
    TreeNode* buildTree(vector<int>& pre, vector<int>& in, int l, int r){
        if(l > r)
            return nullptr;
        int m = mp[pre[id]];
        TreeNode* root = new TreeNode(pre[id++]);
        root->left = buildTree(pre, in, l, m - 1);
        root->right = buildTree(pre, in, m + 1, r);
        return root;
    }
    
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        n = inorder.size();
        id = 0;
        for(int i = 0; i < n; i++)
            mp[inorder[i]] = i;
        return buildTree(preorder, inorder, 0, n - 1);
    }
};

File: July 2022/July_14.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        n = inorder.length;
        id = 0;
        map = new HashMap<>();
        for(int i = 0; i < n; i++)
            map.put(inorder[i], i);
        return buildTree(preorder, inorder, 0, n - 1);
    }
    
    int n, id;
    Map<Integer, Integer> map;
    
    private TreeNode buildTree(int[] pre, int[] in, int l, int r){
        if(l > r)
            return null;
        int m = map.get(pre[id]);
        TreeNode node = new TreeNode(pre[id++]);
        node.left = buildTree(pre, in, l, m - 1);
        node.right = buildTree(pre, in, m + 1, r);
        return node;
    }

}
